src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -163,12 +158,33 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
 
             public void OnResponseHeadersComplete(bool endStream)
             {
-                _responseHeadersAvailable.SetResult(endStream);
+                TaskCompletionSource<bool> waiterTaskSource = null;
+
+                lock (SyncObject)
+                {
+                    if (_state != StreamState.ExpectingHeaders)
+                    {
+                        throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+                    }
+
+                    _state = endStream ? StreamState.Complete : StreamState.ExpectingData;
+
+                    if (_waiterTaskSource != null)
+                    {
+                        waiterTaskSource = _waiterTaskSource;
+                        _waiterTaskSource = null;
+                    }
+                }
+
+                if (waiterTaskSource != null)
```

Comment:
This new pattern is much nicer / simpler / easier to prove is correct from a threading perspective. Thanks.